File: examples/tealeaf/ppcg_calc_rrn_kernel/fpga/ppcg_calc_rrn_kernel.cpp
/***************************************************************
    Copyright 2023 Hewlett Packard Enterprise Development LP.
****************************************************************/

void ppcg_calc_rrn_kernel(int x_min, int x_max, int y_min, int y_max, int halo_exchange_depth,
                          double * r, double * r_store, double * z, double * rrn) {

    int x_dim = x_max - x_min;
    int y_dim = y_max - y_min;

    int r_x  = 2 * halo_exchange_depth + x_dim + 1;

    for(int k = 0; k <= y_dim; k++) {
        for(int j = 0; j <= x_dim; j++) {
            *rrn = *rrn + (r[(k + halo_exchange_depth) * r_x + (halo_exchange_depth + j)] - r_store[(k + halo_exchange_depth) * r_x + (halo_exchange_depth + j)]) * z[(k + halo_exchange_depth) * r_x + (halo_exchange_depth + j)];

        }
    }
}

File: examples/tealeaf/ppcg_calc_rrn_kernel/fpga/ppcg_calc_rrn_kernel.f90
!************************************************************
! Copyright 2023 Hewlett Packard Enterprise Development LP.
!************************************************************

SUBROUTINE ppcg_calc_rrn_kernel(x_min, &
                                         x_max, &
                                         y_min, &
                                         y_max, &
                                         halo_exchange_depth, &
                                         r, &
                                         r_store, &
                                         z, &
                                         rrn )

  IMPLICIT NONE
  INTEGER(KIND=4) :: j,k
  INTEGER(KIND=4) :: x_min, x_max, y_min, y_max, halo_exchange_depth
  REAL(KIND=8) :: rrn
  REAL(KIND=8), DIMENSION(x_min-halo_exchange_depth:x_max+halo_exchange_depth,&
                          y_min-halo_exchange_depth:y_max+halo_exchange_depth) :: r, r_store, z

  rrn = 0.0_8

  DO k=y_min,y_max
    DO j=x_min,x_max
      rrn = rrn + (r(j, k) - r_store(j, k))*z(j, k)
    ENDDO
  ENDDO

END SUBROUTINE ppcg_calc_rrn_kernel
